### PR TITLE
docs: add supabase guide

### LINKED ISSUE #279: [TRI-949] Create a self-hosting guide powered by a Supabase database
Since we use postgres under the hood, it would be very cool if it was possible to use an existing Supabase instance instead of having to spin up a different instance of postgres just for Trigger.dev. The first thing that will need to happen is to do some research to see if it's possible, and if it's not, propose any changes to the project to make it possible. One thing to note is that we make use of two different postgres [schemas](https://www.postgresql.org/docs/current/ddl-schemas.html), one for the main database tables and one for graphile worker (named `graphile_worker` by default). So you'll probably need to figure out a way to specify schemas names for both of those that don't conflict with Supabase built in ones. We also use prisma so you'll need to research to see if it's possible to use that with Supabase.

The guide should be created in the Self Hosting section of the docs:

The doc files are in this monorepo [here](https://github.com/triggerdotdev/trigger.dev/tree/main/docs/documentation/guides/self-hosting) and are written in MDX, powered by Mintlify.

Feel free to just run the docker container locally and not include instructions for how to deploy that to a production service, as that will be covered in other guides (or equally feel free to cover that as well).

<sub>From [SyncLinear.com](https://synclinear.com) | [TRI-949](https://linear.app/triggerdotdev/issue/TRI-949/create-a-self-hosting-guide-powered-by-a-supabase-database)</sub>

### PR DESCRIPTION
**Pull request by nicktrn**
Adds a Supabase guide to the self-hosting section.

WIP for now. Will be adding connection pooling via [PgBouncer](https://www.pgbouncer.org/) to the guide.

This has first class support in both Supabase and Prisma, and should benefit not just self-hosters but the project in general. Can be easily self-hosted, and reduces the need for [Data Proxy](https://www.prisma.io/data-platform/proxy).

There is also a trimmed down compose file in the guide that may benefit from being included in [triggerdotdev/docker](https://github.com/triggerdotdev/docker)

Fixes #279 
/claim #279

### CODE REVIEW

File: docs/documentation/guides/self-hosting/supabase.mdx [hunk 1]
```diff
@@ -0,0 +1,166 @@
+---
+title: "Deploy with Supabase"
+description: "Self-host [Trigger.dev](https://trigger.dev) with Supabase and Docker"
+---
+
+## Requirements
+
+- [docker](https://docs.docker.com/engine/install/)
+- time
```
Comment (ericallam) [inline docs/documentation/guides/self-hosting/supabase.mdx]: 😂

**Comment by ericallam** [comment]
This looks amazing! Great work 👍🏼. One requested tweak: could you specify a separate schema other than `public` for the Trigger.dev tables. Maybe just `triggerdotdev`?

**Comment by nicktrn** [comment]
Sure thing. Will change that when I get a chance to finish up. Just FYI, changing the graphile-worker schema is also possible, should the need ever arise.

**Comment by ericallam** [comment]
@nicktrn would it be possible to get this merged in before 3pm GMT today? We're announcing something at 4pm today in a blog post and would love to be able to link to it 👍

**Comment by nicktrn** [comment]
@ericallam Not a problem, :coffee: solves (almost) anything

**Comment by nicktrn** [comment]
@ericallam I have added support for connection pools. Tests are passing. Will finish the writeup now. If you could have a look at this immediately, that would save time.

Worst case, will have to publish without the bonus section on connection pools.

**Comment by ericallam** [comment]
Oh shoot I didn't realize this had changes to the webapp to get it to work. Unfortunately, we probably won't be able to get it in under the deadline as I don't have time to pull it down and test. So I think I will link to this PR from the blog post instead. Sorry again and thanks for this! Will hopefully be able to get it merged tomorrow 👍

**Comment by nicktrn** [comment]
No worries, would you like me to take out the last section for now and revert the webapp changes?

That way you can publish the main part of the guide without any trouble and link to it directly.

**Comment by nicktrn** [comment]
@ericallam linked PR without app changes. Hope this helps!

- #299

**Comment by nicktrn** [comment]
@ericallam this is ready for review now.

End-result is that _nothing_ changes unless `DIRECT_URL` is manually set.

The tests are passing fine manually with and without pooling. Managed to get it working via playwright, but quite flaky.

There are issues with the tests in general. Boils down to responses not being awaited before navigating, timeouts not considering latency (dev/prod, local/remote db), and setup/teardown being a mess. Ideally this should be done with fixtures or dependencies instead. Will have a proper look when I get a chance and sort things out.

Re the `.env` symlink, can avoid that for now if you'd like to carry on with the previous pattern. Just need to add a `DIRECT_URL=${DATABASE_URL}` line to set a default, just as I've done in `.env.example`

File: .env.example [hunk 2]
```diff
@@ -3,7 +3,10 @@ SESSION_SECRET=abcdef1234
 MAGIC_LINK_SECRET=abcdef1234
 ENCRYPTION_KEY=ae13021afef0819c3a307ad487071c06 # Must be a random 16 byte hex string. You can generate an encryption key by running `openssl rand -hex 16` in your terminal
 LOGIN_ORIGIN=http://localhost:3030
-DATABASE_URL=postgresql://postgres:postgres@localhost:5432/postgres?schema=public 
+DATABASE_URL=postgresql://postgres:postgres@localhost:5432/postgres?schema=public
+# This sets the URL used for direct connections to the database and should only be needed in limited circumstances
+# See: https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#fields:~:text=the%20shadow%20database.-,directUrl,-No
+DIRECT_URL=${DATABASE_URL}
```
Comment (nicktrn) [inline .env.example:9]: This can't just be empty as we have to use an env var in schema.prisma

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -122,7 +122,6 @@ jobs:
           # Setup environment variables
           cp ./.env.example ./.env
           cp ./examples/nextjs-test/.env.example ./examples/nextjs-test/.env.local
-          cp ./packages/database/.env.example ./packages/database/.env
```
Comment (nicktrn) [inline .github/workflows/publish.yml:125]: not needed anymore due to symlink

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -37,11 +37,11 @@ branch are tagged into a release monthly.
    ```
    pnpm i
    ```
-4. Create your `.env` files
+4. Create your `.env` file
    ```
-   cp .env.example .env && cp packages/database/.env.example packages/database/.env
+   cp .env.example .env
```
Comment (nicktrn) [inline CONTRIBUTING.md:42]: not needed anymore due to symlink

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -178,7 +178,6 @@ To run the end-to-end tests, follow the steps below:
 ```sh
 cp ./.env.example ./.env
 cp ./examples/nextjs-test/.env.example ./examples/nextjs-test/.env.local
-cp ./packages/database/.env.example ./packages/database/.env
```
Comment (nicktrn) [inline CONTRIBUTING.md:181]: not needed anymore due to symlink

File: apps/webapp/app/services/worker.server.ts [hunk 6]
```diff
@@ -94,6 +94,7 @@ function getWorkerQueue() {
       connectionString: env.DATABASE_URL,
       concurrency: 5,
       pollInterval: 1000,
+      noPreparedStatements: env.DATABASE_URL !== env.DIRECT_URL,
```
Comment (nicktrn) [inline apps/webapp/app/services/worker.server.ts:97]: No way around this with pooling

**Comment by ericallam** [comment]
I'm happy with the `.env` symlink thing (which I've seen done elsewhere as well). Can you help me understand the need for the DIRECT_URL env var? Is there any way to make it an optional env var or do we have to require it?

**Comment by nicktrn** [comment]
Sure. The Prisma CLI requires a direct connection for [some commands](https://www.prisma.io/docs/data-platform/classic-projects/data-proxy/prisma-cli-with-data-proxy#prisma-cli-commands-that-require-a-direct-database-connection), i.e. not via an external pool. Unless you hardcode this in the schema file, the only way is to use an env var.

Sadly, once we do this, that env var can't be unset / empty as Prisma doesn't allow us to set defaults and _will_ complain otherwise:

```sh
# unset
error: Environment variable not found: DIRECT_URL.
  -->  schema.prisma:4

# empty
error: Error validating datasource `db`: You must provide a nonempty direct URL. The environment variable `DIRECT_URL` resolved to an empty string.
  -->  schema.prisma:4
```

**Comment by ericallam** [comment]
That makes sense. Let me know when this is ready for review 👍

**Comment by nicktrn** [comment]
Should be ready now!

**Comment by ericallam** [comment]
Okay, I just pulled this down and took a look at this and it all looks great. There is one issue though, our self-hosting guides will need to be updated:

- [ ] https://trigger.dev/docs/documentation/guides/self-hosting/flyio
- [ ] https://trigger.dev/docs/documentation/guides/self-hosting/render

And our docker repo:

https://github.com/triggerdotdev/docker

**Comment by nicktrn** [comment]
Yes, I did think that as well. I'll have a look and see what needs changing!

**Comment by nicktrn** [comment]
Fixed the Fly.io guide, got an error following the instructions however. Creating the db failed as the default instance couldn't run Postgres:

```sh
Select configuration: Development - Single node, 1x shared CPU, 256MB RAM, 1GB disk
Failed creating the Postgres cluster triggerdotdev-db: failed to launch VM: aborted: could not reserve resource for machine: insufficient memory available to fulfill request
```

The next tier up was shared-cpu-2x - CPU Kind: Shared, vCPUs: 2, Memory: 4GB which worked. Why the jump in specs I wonder..

The Render guide is a different matter. Don't know what to do there. It doesn't seem to have a DB attached and I couldn't set up an instance following those steps. My guess is it could all be configured in `render.yml` directly, including the DB and pooling (!). See [here](https://render.com/docs/databases#use-renderyaml-for-deploys).

Made a quick change to the docker repo as well and small README change to fly.io to follow shortly.

**Comment by ericallam** [comment]
This looks good, I'm going to merge this and the other PRs now and create a new `latest` release of the docker image.

**Comment by nicktrn** [comment]
Tracking this here in case [Postgres Extensions](https://www.prisma.io/docs/concepts/components/prisma-schema/postgresql-extensions#how-to-enable-the-postgresqlextensions-preview-feature) are ever added to the Prisma schema:
- https://github.com/prisma/prisma/issues/19100

**Pull request merged**